fuzzing/reversion/revert_report.go

```diff
@@ -0,0 +1,163 @@
+	return &RevertReport{
+		RevertedCallReasons: make(map[functionSelector]map[errorSelector]uint),
+		RevertedCalls:       make(map[functionSelector]uint),
+		TotalCalls:          make(map[functionSelector]uint),
+	}
+}
+
+// initFuncSelectorIfMissing is a utility function used to initialize RevertReport for a specific function selector
+func (s *RevertReport) initFuncSelectorIfMissing(selector functionSelector) {
+	_, ok := s.TotalCalls[selector]
+	if !ok {
+		s.TotalCalls[selector] = 0
+		s.RevertedCalls[selector] = 0
+		s.RevertedCallReasons[selector] = make(map[errorSelector]uint)
+	}
+}
+
+// initErrSelectorIfMissing is a utility function used to initialize RevertedCallReasons for a specific error selector
+func (s *RevertReport) initErrSelectorIfMissing(funcSelector functionSelector, errSelector errorSelector) {
+	_, ok := s.RevertedCallReasons[funcSelector][errSelector]
+	if !ok {
+		s.RevertedCallReasons[funcSelector][errSelector] = 0
+	}
+}
+
+// addCall is used to add a single call to the revert report
+func (s *RevertReport) addCall(funcSelector functionSelector, errSelector errorSelector, didRevert bool) {
+	s.addCallCount(funcSelector, errSelector, didRevert, 1)
+}
+
+// addCallCount is used to add one or more calls to the revert report.
+func (s *RevertReport) addCallCount(funcSelector functionSelector, errSelector errorSelector, didRevert bool, number uint) {
+	s.initFuncSelectorIfMissing(funcSelector)
+	s.TotalCalls[funcSelector] += number
+	if didRevert {
+		s.RevertedCalls[funcSelector] += number
+		s.initErrSelectorIfMissing(funcSelector, errSelector)
+		s.RevertedCallReasons[funcSelector][errSelector] += number
+	}
+}
+
+// concatReports Subsumes the data from the `other` report into the receiver report.
+func (s *RevertReport) concatReports(other *RevertReport) {
+	for fSel, callCount := range other.TotalCalls {
+		revertCount := other.RevertedCalls[fSel]
+		revertReasons := other.RevertedCallReasons[fSel]
+
+		for revReason, revCount := range revertReasons {
+			s.addCallCount(fSel, revReason, true, revCount)
+		}
+
+		successCount := callCount - revertCount
+		s.addCallCount(fSel, errorSelector{}, false, successCount)
+	}
+}
+
+// ToArtifact Converts the revert report to an artifact object. Does not populate previous report data or sort the data.
+func (s *RevertReport) ToArtifact(contractDefs fuzzerTypes.Contracts) *ReportArtifact {
+	funcLookup, errLookup := buildSelectorLookups(contractDefs)
+	artifact := &ReportArtifact{}
+
+	for funcSel, runCount := range s.TotalCalls {
+		revertCount := s.RevertedCalls[funcSel]
+		funcName := funcLookup[funcSel]
+		funcArtifact := &FunctionArtifact{
+			Name:          funcName,
+			TotalCalls:    runCount,
+			TotalReverts:  revertCount,
+			RevertPct:     float64(revertCount) / float64(runCount),
+			PrevRevertPct: nil,
+			RevertReasons: []*RevertReasonArtifact{},
+		}
+
+		for errSel, errCount := range s.RevertedCallReasons[funcSel] {
+			errName, ok := errLookup[errSel]
```

Comment:
One issue with this approach versus a tracer is that we don't actually know the codehash of the contract that created the revert data and won't handle collisions correctly. Ideally we could use https://github.com/crytic/medusa/blob/master/compilation/abiutils/solidity_errors.go#L90

Reply:
I think it's okay that we eat that risk since the current approach gives us the following benefits:
1. Tracking reversion stats using the current method has very little overhead. I ran medusa for an hour with the feature both enabled and disabled and the call count difference was <5%. Using a tracer would probably be a lot more impactful to performance.
2. A function selector collision in the entrypoint contract would get flagged by the compiler, so a theoretical function selector collision could only happen in multi-abi mode which nobody uses anyways.
3. If there's an error selector collision, the developer has bigger problems to worry about because this means their off-chain components will never be able to handle errors properly. 
4. I really don't want to rewrite the code ;)

Reply:
I'm fine with leaving it as is, but it's worth documenting. The error that causes the revert in the top level call isn't necessarily defined in that contract as often times the revert is bubbled up from other call frames.

Reply:
I'll add a note to the report header. 

For all we know, after using these reports for another invar engagement, we'll want to figure out the proper source of the revert anyways and end up implementing it as a tracer